llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp

```diff
@@ -676,7 +676,14 @@ Value *SCEVExpander::visitUDivExpr(const SCEVUDivExpr *S) {
                          SCEV::FlagAnyWrap, /*IsSafeToHoist*/ true);
   }
 
-  Value *RHS = expand(S->getRHS());
+  const SCEV *RHSExpr = S->getRHS();
+  Value *RHS = expand(RHSExpr);
+  if (SafeUDivMode && !SE.isKnownNonZero(RHSExpr)) {
```

Comment:
isKnownNonZero doesn't check for poison, I think.

Reply:
Hm, let me see if I can come up with a case where it is known-non-zero, but may be poison

Reply:
Something like umax(x, 1) should do the trick.

Reply:
Added and updated to separately check for !poison -> emit freeze, !known-non-zero -> emit umax